### PR TITLE
reviewing metasyncer code

### PR DESCRIPTION
**Pull request by usmanong**
Branch based on metasync merged with master

### CODE REVIEW

File: dfc/metasync.go [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+		glog.Warningf("... failing to sync %s, err: %v (%d)", si.DaemonID, err, status)
+		return
+	}
+	y.pending.lock.Lock()
+	delete(y.pending.diamonds, si.DaemonID)
+	y.pending.lock.Unlock()
+}
+
+func (y *metasyncer) handleRefused(urlfmt, method string, jsbytes []byte, smap4bcast *Smap) {
+	wg := &sync.WaitGroup{}
```
Comment (VladimirMarkelov) [inline dfc/metasync.go]: wg.Wait() missing?

File: dfc/daemon.go [hunk 2]
```diff
@@ -163,39 +160,68 @@ func (m *Smap) count() int {
...
+func (m *Smap) countL() int {
+	smapLock.Lock()
+	defer smapLock.Unlock()
 	return m.count()
 }
 
 func (m *Smap) get(sid string) *daemonInfo {
-	si := m.Tmap[sid]
+	si, ok := m.Tmap[sid]
+	if !ok {
```
Comment (liangdrew) [inline dfc/daemon.go]: This function can probably just be one line - i.e. `return m.Tmap[sid]`. See https://play.golang.org/p/cncYTcOVeyN. Same for `getProxy()` below.

File: dfc/proxy.go [hunk 3]
```diff
@@ -608,7 +598,23 @@ func (p *proxyrunner) httpbckdelete(w http.ResponseWriter, r *http.Request) {
...
+		lbmapLock.Lock()
+		defer lbmapLock.Unlock()
+		if !p.lbmap.del(bucket) {
+			s := fmt.Sprintf("Local bucket %s does not exist", bucket)
+			p.invalmsghdlr(w, r, s)
+			return
+		}
+		if errstr := p.savelbmapconf(); errstr != "" {
+			p.lbmap.add(bucket)
+			p.invalmsghdlr(w, r, errstr)
```
Comment (VladimirMarkelov) [inline dfc/proxy.go]: return missing?
The similar code below for create LB does 'return' in this case.

File: dfc/proxy.go [hunk 4]
```diff
@@ -1747,17 +1689,48 @@ func (p *proxyrunner) httpclupost(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+	if errstr := p.savesmapconf(); errstr != "" {
+		// take it back
+		if proxy {
+			p.smap.delProxy(nsi.DaemonID)
+		} else {
+			p.smap.del(nsi.DaemonID)
+		}
+		p.invalmsghdlr(w, r, errstr)
```
Comment (VladimirMarkelov) [inline dfc/proxy.go]: Do we need return here?

File: dfc/metasync.go [hunk 5]
```diff
@@ -0,0 +1,719 @@
...
+func (y *metasyncer) dosync(revsvec []interface{}) int {
+	var (
+		smap4bcast, smapSynced *Smap
+		jsbytes, jsmsg         []byte
+		err                    error
+		payload                = make(map[string]string)
+		newversions            = make(map[string]revs)
+		check4newmembers       bool
+	)
+	if _, ok := y.synced.copies[smaptag]; ok {
```
[the remark this answers is not in the document]
Comment (VladimirMarkelov) [reply dfc/metasync.go]: I saw the similar double-get-from-hash way in a few other places.

**Pull request merged**